addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,479 @@
+            this.fixedIp = true;
+        }
+
+        // When we set up this object, we also want to get the proper gson adapter set up once
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(NhcMessageBase.class, new NikoHomeControlMessageDeserializer());
+        this.gsonIn = gsonBuilder.create();
+    }
+
+    public NikoHomeControlCommunication() throws IOException {
+
+        this(null, null);
+    }
+
+    /**
+     * Start communication with Niko Home Control IP-interface, run through initialization and start thread listening
+     * to all messages coming from Niko Home Control.
+     *
+     */
+    public void startCommunication(int port) {
+
+        try {
+            for (int i = 1; nhcEventsRunning && (i <= 5); i++) {
+                // the events listener thread did not finish yet, so wait max 5000ms before restarting
+                Thread.sleep(1000);
+            }
+            if (nhcEventsRunning) {
+                logger.error(
+                        "Niko Home Control: starting from thread {}, but previous connection still active after 5000ms",
+                        Thread.currentThread().getId());
+                throw new IOException();
+            }
+
+            this.nhcPort = port;
+            this.nhcSocket = new Socket(this.nhcAddress, this.nhcPort);
+            this.nhcOut = new PrintWriter(this.nhcSocket.getOutputStream(), true);
+            this.nhcIn = new BufferedReader(new InputStreamReader(this.nhcSocket.getInputStream()));
+            logger.info("Niko Home Control: connected from thread {}", Thread.currentThread().getId());
+
+            // initialize all info in local fields
+            initialize();
+
+            // Start Niko Home Control event listener. This listener will act on all messages coming from
+            // IP-interface.
+            (new Thread(nhcEvents)).start();
+
+        } catch (IOException | InterruptedException e) {
+            // if the error occurs in the initialization, don't try to restart
+            logger.warn("Niko Home Control: error initializing communication from thread {}",
+                    Thread.currentThread().getId());
+            stopCommunication();
+            // take bridge offline
+            this.bridgeCallBack.bridgeOffline();
+        }
+
+    }
+
+    /**
+     * Cleanup socket when the communication with Niko Home Control IP-interface is closed.
+     *
+     */
+    public synchronized void stopCommunication() {
+        this.listenerStopped = true;
+
+        if (this.nhcSocket != null) {
+            try {
+                this.nhcSocket.close();
+            } catch (IOException ignore) {
+                // ignore IO Error when trying to close the socket if the intention is to close it anyway
+            }
+            this.nhcSocket = null;
+        }
+        logger.warn("Niko Home Control: communication stopped from thread {}", Thread.currentThread().getId());
```

Comment:
Shouldn't this be debug, because it happens at any regular shutdown of the handler?